registry/storage/blobstore.go

```diff
@@ -90,6 +111,130 @@ func (bs *blobStore) Put(ctx context.Context, mediaType string, p []byte) (distr
 	}, bs.driver.PutContent(ctx, bp, p)
 }
 
+func (bs *blobStore) Enumerate(ctx context.Context, digests []digest.Digest, last string) (n int, err error) {
+	var lastPathComponents []string
+
+	if last != "" {
+		lastPathComponents, err = digestPathComponents(digest.Digest(last), true)
+		if err != nil {
+			return 0, fmt.Errorf("invalid last digest %q", last)
+		}
+	}
+
+	rootPath := path.Join(storagePathRoot, storagePathVersion, "blobs")
+
+	pastLastOffset := func(pthSepCount int, pth string, lastPathComponents []string) bool {
+		maxComponents := pthSepCount + 1
+		if maxComponents >= len(lastPathComponents) {
+			return pth > path.Join(lastPathComponents...)
+		}
+		return pth >= path.Join(lastPathComponents[:maxComponents]...)
+	}
+
+	err = WalkSorted(ctx, bs.driver, rootPath, func(fi driver.FileInfo) error {
+		if !fi.IsDir() {
+			// ignore files
+			return nil
+		}
+
+		// trim <from>/ prefix
+		pth := strings.TrimPrefix(strings.TrimPrefix(fi.Path(), rootPath), "/")
+		sepCount := strings.Count(pth, "/")
+
+		if !pastLastOffset(sepCount, pth, lastPathComponents) {
+			// we haven't reached the 'last' offset yet
+			return ErrSkipDir
+		}
+
+		if sepCount < 2 {
+			// continue walking
+			return nil
+		}
+
+		alg := ""
+		tarsumParts := reTarsumPrefix.FindStringSubmatch(pth)
+		isTarsum := len(tarsumParts) > 0
+		if sepCount > 4 || (!isTarsum && sepCount > 2) {
```

Comment:
I'd rather move this walk function into a factory where these value will be defined based on factory's parameters. Its functionality is mostly shared with `linkedBlobStore.Enumerate()` therefor will keep it DRY. For now I'll move it into `registry/storage/walk.go`.